public/app/features/query/state/QueryRunner.ts

```diff
@@ -0,0 +1,115 @@
+import {
+  CoreApp,
+  DataQueryRequest,
+  DataSourceApi,
+  PanelData,
+  rangeUtil,
+  ScopedVars,
+  QueryRunnerOptions,
+  QueryRunner as QueryRunnerSrv,
+} from '@grafana/data';
+import { getTemplateSrv } from '@grafana/runtime';
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { cloneDeep } from 'lodash';
+import { Observable, ReplaySubject, Unsubscribable } from 'rxjs';
+import { getNextRequestId } from './PanelQueryRunner';
+import { preProcessPanelData, runRequest } from './runRequest';
+export class QueryRunner implements QueryRunnerSrv {
+  private subject: ReplaySubject<PanelData>;
+  private subscription?: Unsubscribable;
+  private lastResult?: PanelData;
+
+  constructor() {
+    this.subject = new ReplaySubject(1);
+  }
+
+  get(): Observable<PanelData> {
+    return this.subject;
+  }
+
+  async run(options: QueryRunnerOptions): Promise<void> {
+    const {
+      queries,
+      timezone,
+      datasource,
+      panelId,
+      app,
+      dashboardId,
+      timeRange,
+      timeInfo,
+      cacheTimeout,
+      maxDataPoints,
+      scopedVars,
+      minInterval,
+    } = options;
+
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+    }
+
+    const request: DataQueryRequest = {
+      app: app ?? CoreApp.Unknown,
+      requestId: getNextRequestId(),
+      timezone,
+      panelId,
+      dashboardId,
+      range: timeRange,
+      timeInfo,
+      interval: '',
+      intervalMs: 0,
+      targets: cloneDeep(queries),
+      maxDataPoints: maxDataPoints,
+      scopedVars: scopedVars || {},
+      cacheTimeout,
+      startTime: Date.now(),
+    };
+
+    // Add deprecated property
+    (request as any).rangeRaw = timeRange.raw;
+
+    try {
+      const ds = await getDataSource(datasource, request.scopedVars);
+
+      // Attach the datasource name to each query
+      request.targets = request.targets.map((query) => {
+        if (!query.datasource) {
+          query.datasource = ds.name;
+        }
+        return query;
+      });
+
+      const lowerIntervalLimit = minInterval ? getTemplateSrv().replace(minInterval, request.scopedVars) : ds.interval;
+      const norm = rangeUtil.calculateInterval(timeRange, maxDataPoints, lowerIntervalLimit);
+
+      // make shallow copy of scoped vars,
+      // and add built in variables interval and interval_ms
+      request.scopedVars = Object.assign({}, request.scopedVars, {
+        __interval: { text: norm.interval, value: norm.interval },
+        __interval_ms: { text: norm.intervalMs.toString(), value: norm.intervalMs },
+      });
+
+      request.interval = norm.interval;
+      request.intervalMs = norm.intervalMs;
+
+      this.subscription = runRequest(ds, request).subscribe({
+        next: (data) => {
```

Comment:
nit: should we add `error: (err) => {...}` for the error handling or is this caught somewhere else? Also, maybe add `complete` that could unsubscribe the subscription?

Reply:
Yea, was surprised that his looked like this in PanelQueryRunner, should an error be bubbled to the subject=? 

does not complete unsubscribe automatically?

Reply:
Great input @hugohaggmark! Adding the improvements.

Reply:
Hmm, that's a good question, you're probably right that the complete will call unsubscribe. Nevermind that. Regarding the error, I think that bubbles up to the subject; I was mostly wondering if we should make it explicit by handling it here.